File: packages/python/awaithumans/types/verification.py
"""Verification types — verifier config, context, and result."""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field


class VerificationContext(BaseModel):
    """Context passed to the verifier for quality checking + NL parsing."""

    task: str
    payload: Any
    payload_schema: dict  # JSON Schema
    response: Any | None = None  # structured response (None if NL input)
    response_schema: dict  # JSON Schema
    raw_input: str | None = None  # natural language text
    attempt: int
    previous_rejections: list[str] = Field(default_factory=list)


class VerifierResult(BaseModel):
    """Result from the verifier."""

    passed: bool
    reason: str = Field(description="Human-readable — shown to the human if rejected.")
    parsed_response: Any | None = Field(
        default=None,
        description="Extracted from NL input, conforming to response_schema.",
    )


class VerifierConfig(BaseModel):
    """Configuration for a verifier (passed to the server, executed server-side)."""

    provider: str = Field(description='E.g., "claude", "openai", "gemini", or "custom"')
    model: str | None = Field(default=None, description='E.g., "claude-sonnet-4-20250514"')
    instructions: str = Field(description="Prompt template for the verifier.")
    max_attempts: int = Field(default=3, ge=1, le=10)
    api_key_env: str | None = Field(
        default=None,
        description="Env var name for the API key. Server reads this at runtime.",
    )
    metadata: dict[str, Any] | None = Field(
        default=None,
        description="Provider-specific config (e.g., Azure endpoint_env, api_version).",
    )

File: packages/typescript-sdk/src/types/verification.ts
/**
 * Verifier types — config the SDK passes to the server and reference
 * shapes the server's verifier implementations consume / return.
 *
 * Mirrors `awaithumans/types/verification.py`. Verification runs
 * server-side; the SDK just hands across the config blob.
 */

import type { JsonSchema7Type } from "zod-to-json-schema";

/**
 * Verifier configuration — sent to the server, which executes it.
 * The SDK does NOT run verification locally. This is a config blob.
 */
export interface VerifierConfig {
	provider: string;
	model?: string;
	instructions: string;
	maxAttempts: number;
	apiKeyEnv?: string;
}

/**
 * Server-side verifier interface. Exported for reference / adapter
 * authors. The Python server implements this, not the TS SDK.
 */
export interface Verifier {
	verify(context: VerificationContext): Promise<VerifierResult>;
	maxAttempts: number;
}

export interface VerificationContext {
	task: string;
	payload: unknown;
	payloadSchema: JsonSchema7Type;
	/** Structured response from the human. Null if NL input. */
	response: unknown | null;
	responseSchema: JsonSchema7Type;
	/** Natural language text from Slack thread / email reply. */
	rawInput?: string;
	/** Which verification attempt (1, 2, 3...). */
	attempt: number;
	/** Reasons from prior failed verification attempts. */
	previousRejections: string[];
}

export interface VerifierResult {
	passed: boolean;
	/** Human-readable — shown to the human if rejected. */
	reason: string;
	/** Extracted from NL input, conforming to responseSchema. Only present when parsing NL. */
	parsedResponse?: unknown;
}
